Inconsistent Request Body Caching Between c.req.json() and Validator Middleware

### What version of Hono are you using?

3.12.11

### What runtime/platform is your app running on?

Cloudflare Workers

### What steps can reproduce the bug?

I've encountered an issue with how Hono caches the request body in different formats, leading to unexpected behavior when using middlewares in a specific order. 

When one middleware calls `c.req.json()`, it updates `bodyCache[json]` without also updating `bodyCache[arrayBuffer]`. 
If we later use the `validator` middleware, which expects the `arrayBuffer` to be cached, it results in an error for a malformed request.

```ts
app.post(
  '/test',
  async (c, next) => {
    const body = await c.req.json();
    await next();
  },
  validator('json', (value, c) => {
    return value;
  }),
  async (c) => {
    return c.json({ message: 'OK' }, 200);
  },
);
```

### What is the expected behavior?

The request passes through both middlewares without issues, with the validator correctly validating the parsed JSON.

### What do you see instead?

Receive a 400 Bad Request error.
```
[  <-- POST /TEST]: undefined
✘ [ERROR] Error: Malformed JSON in request body
```

### Additional information

The validator middleware throws an error indicating a malformed request body, despite the body being valid JSON. This seems to occur because c.req.json() does not update bodyCache[arrayBuffer], leading to a mismatch in the cached body data.

```ts
// request.ts
  private cachedBody = (key: keyof Body) => {
// ...
    if (bodyCache.arrayBuffer) {
      return (async () => {
        return await new Response(bodyCache.arrayBuffer)[key]()
      })()
    }
    return (bodyCache[key] = raw[key]()) // doesn't update arrayBuffer
  }
```

```ts
// validator/validator.ts
        /**
         * Get the arrayBuffer first, create JSON object via Response,
         * and cache the arrayBuffer in the c.req.bodyCache.
         */
        try {
          const arrayBuffer = c.req.bodyCache.arrayBuffer ?? (await c.req.raw.arrayBuffer()) // Uses arrayBuffer
          value = await new Response(arrayBuffer).json()
          c.req.bodyCache.json = value
          c.req.bodyCache.arrayBuffer = arrayBuffer
        } catch {
          console.error('Error: Malformed JSON in request body')
          return c.json(
            {
              success: false,
              message: 'Malformed JSON in request body',
            },
            400
          )
        }
```

--- a/deno_dist/validator/validator.ts
+++ b/deno_dist/validator/validator.ts
@@ -67,9 +67,12 @@ export const validator = <
           const message = `Invalid HTTP header: Content-Type=${contentType}`
           throw new HTTPException(400, { message })
         }
+        if (c.req.bodyCache.json) {
+          value = await c.req.bodyCache.json
+          break
+        }
         /**
-         * Get the arrayBuffer first, create JSON object via Response,
-         * and cache the arrayBuffer in the c.req.bodyCache.
+         * Get the arrayBuffer, create JSON object via Response, and cache the arrayBuffer in the c.req.bodyCache if there is no cached content in c.req.bodyCache.json.
          */
         try {
           const arrayBuffer = c.req.bodyCache.arrayBuffer ?? (await c.req.raw.arrayBuffer())
@@ -82,6 +85,10 @@ export const validator = <
         }
         break
       case 'form': {
+        if (c.req.bodyCache.formData) {
+          value = c.req.bodyCache.formData
+          break
+        }
         try {
           const contentType = c.req.header('Content-Type')
           if (contentType) {

--- a/src/validator/validator.test.ts
+++ b/src/validator/validator.test.ts
@@ -172,6 +172,62 @@ describe('Malformed FormData request', () => {
   })
 })
 
+describe('Cached contents', () => {
+  const app = new Hono()
+
+  app.post(
+    '/json',
+    async (c, next) => {
+      await c.req.json()
+      await next()
+    },
+    validator('json', (value) => {
+      return value
+    }),
+    async (c) => {
+      const data = await c.req.json()
+      return c.json(data, 200)
+    }
+  )
+
+  app.post(
+    '/form',
+    async (c, next) => {
+      await c.req.formData()
+      await next()
+    },
+    validator('form', (value) => {
+      return value
+    }),
+    async (c) => {
+      return c.json({ message: 'OK' }, 200)
+    }
+  )
+
+  it('Should return the cached JSON content', async () => {
+    const res = await app.request('/json', {
+      method: 'POST',
+      body: JSON.stringify({ message: 'Hello' }),
+      headers: {
+        'Content-Type': 'application/json',
+      },
+    })
+    expect(res.status).toBe(200)
+    expect(await res.json()).toEqual({ message: 'Hello' })
+  })
+
+  it('Should return the cached form content', async () => {
+    const form = new FormData()
+    form.append('message', 'Hello')
+    const res = await app.request('/form', {
+      method: 'POST',
+      body: form,
+    })
+    expect(res.status).toBe(200)
+    expect(await res.json()).toEqual({ message: 'OK' })
+  })
+})
+
 describe('Validator middleware with a custom validation function', () => {
   const app = new Hono()
 

--- a/src/validator/validator.ts
+++ b/src/validator/validator.ts
@@ -67,9 +67,12 @@ export const validator = <
           const message = `Invalid HTTP header: Content-Type=${contentType}`
           throw new HTTPException(400, { message })
         }
+        if (c.req.bodyCache.json) {
+          value = await c.req.bodyCache.json
+          break
+        }
         /**
-         * Get the arrayBuffer first, create JSON object via Response,
-         * and cache the arrayBuffer in the c.req.bodyCache.
+         * Get the arrayBuffer, create JSON object via Response, and cache the arrayBuffer in the c.req.bodyCache if there is no cached content in c.req.bodyCache.json.
          */
         try {
           const arrayBuffer = c.req.bodyCache.arrayBuffer ?? (await c.req.raw.arrayBuffer())
@@ -82,6 +85,10 @@ export const validator = <
         }
         break
       case 'form': {
+        if (c.req.bodyCache.formData) {
+          value = c.req.bodyCache.formData
+          break
+        }
         try {
           const contentType = c.req.header('Content-Type')
           if (contentType) {

